#Library importing

In [ ]:
import numpy as np
import pandas as pd
from keras.utils import to_categorical

#Dataset Importing


In [ ]:
dataset = pd.read_csv('/content/Iris.csv')

In [ ]:
dataset

Id  SepalLengthCm  ...  PetalWidthCm         Species
0      1            5.1  ...           0.2     Iris-setosa
1      2            4.9  ...           0.2     Iris-setosa
2      3            4.7  ...           0.2     Iris-setosa
3      4            4.6  ...           0.2     Iris-setosa
4      5            5.0  ...           0.2     Iris-setosa
..   ...            ...  ...           ...             ...
145  146            6.7  ...           2.3  Iris-virginica
146  147            6.3  ...           1.9  Iris-virginica
147  148            6.5  ...           2.0  Iris-virginica
148  149            6.2  ...           2.3  Iris-virginica
149  150            5.9  ...           1.8  Iris-virginica

[150 rows x 6 columns]

# Dataset PreProcessing

Dataset divide into x and y 

LableEncoder is use for label encoding means I is used for converting String label into integer label.

In [ ]:
x = dataset.iloc[:,1:5].values
y = dataset.iloc[:,5].values

from sklearn.preprocessing import LabelEncoder

encoder = LabelEncoder()
y1 = encoder.fit_transform(y)
#get_dummies is used for convert label into one-hot encoding
y = pd.get_dummies(y1).values
#print(y)

From importing train_test_split we can split our dataset into training part and testing part .

In [ ]:
from sklearn.model_selection import train_test_split

X_train,X_test,y_train,y_test = train_test_split(x,y,test_size=0.2,random_state=0)
#print(X_train)

#Model Building

In [ ]:
from keras.models import Sequential
from keras.layers import Dense
from keras.optimizers import SGD,Adam

model = Sequential()

model.add(Dense(10,input_shape=(4,),activation='relu'))
model.add(Dense(8,activation='relu'))
#model.add(Dense(7,activation='relu'))
model.add(Dense(6,activation='relu'))
model.add(Dense(3,activation='softmax'))

model.compile(Adam(lr=0.1),'categorical_crossentropy',metrics=['accuracy'])

model.summary()


Model: "sequential_10"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_40 (Dense)             (None, 10)                50        
_________________________________________________________________
dense_41 (Dense)             (None, 8)                 88        
_________________________________________________________________
dense_42 (Dense)             (None, 6)                 54        
_________________________________________________________________
dense_43 (Dense)             (None, 3)                 21        
Total params: 213
Trainable params: 213
Non-trainable params: 0
_________________________________________________________________


# Model Training and Evaluating For test dataset

In [ ]:
model.fit(X_train,y_train,epochs=100)
y_prediction = model.predict(X_test)

Epoch 1/100
4/4 [==============================] - 0s 2ms/step - loss: 1.3362 - accuracy: 0.3500
Epoch 2/100
4/4 [==============================] - 0s 2ms/step - loss: 1.1247 - accuracy: 0.3250
Epoch 3/100
4/4 [==============================] - 0s 2ms/step - loss: 1.1434 - accuracy: 0.3250
Epoch 4/100
4/4 [==============================] - 0s 2ms/step - loss: 0.9784 - accuracy: 0.5167
Epoch 5/100
4/4 [==============================] - 0s 2ms/step - loss: 0.6829 - accuracy: 0.6917
Epoch 6/100
4/4 [==============================] - 0s 2ms/step - loss: 0.4104 - accuracy: 0.8917
Epoch 7/100
4/4 [==============================] - 0s 2ms/step - loss: 0.3180 - accuracy: 0.8417
Epoch 8/100
4/4 [==============================] - 0s 2ms/step - loss: 0.3530 - accuracy: 0.8000
Epoch 9/100
4/4 [==============================] - 0s 2ms/step - loss: 0.2094 - accuracy: 0.9250
Epoch 10/100
4/4 [==============================] - 0s 2ms/step - loss: 0.1458 - accuracy: 0.9500
Epoch 11/100
4/4 [===========

# Conclusion:

If we use Sigmid Function in every Layer Then We get result :

loss: 1.0959 - accuracy: 0.3667

Now If we use relu in (L-1) layer and sigmoid in last layer then:
 loss: 0.2717 - accuracy: 0.9167
 loss: 0.4588 - accuracy: 0.7250

 So accuracy is about 70 to 90%

 Now if We use Softmax in last layer:
  loss: 0.1202 - accuracy: 0.9583
  loss: 0.1114 - accuracy: 0.9750
  Accuracy Increased:

  As Relu We can use tanh in every (L-1) layer so accuracy is also increased.


  Learning Rate:

  If we initialize lr=1 then Result is decreased:
  loss: 1.1147 - accuracy: 0.3250

  now lr=0.1 then result:
  loss: 0.1404 - accuracy: 0.9333

  So Accuaracy is highly increased by only change in learning rate.

  As Learning rate  decreased If we increased number of epochs The Accuracy is also increased.


  If We change optimizers to SGD to Adam We can see Our dataset Accuracy is increased in few epochs as this dataset is small so we can't see difference but we can difference in large dataset.

